### PR TITLE
closes #3

### LINKED ISSUE #3: Refactor the Discord bot to use Discord's command setup
Not too long ago, Discord added some really nice UI and discoverability to slash-commands. Unfortunately, it requires a code change on the bot side to support them.

Once #2 is done, we can refactor the Discord bot to use Discord's slash-command system.

### PR DESCRIPTION
**Pull request by reticivis-net**
closes #3
I haven't done end-to-end testing yet, but everything seems to work fine on my end
also PyCharm formatted some files on its own so oops

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -1,107 +1,126 @@
...
+                await lib.benchmark(*submit_msg)
                 self.queue.task_done()
             except Exception:
                 logger.exception("Error while processing submission.")
 
-    async def handle_best(self, msg):
+
+def today():
+    dt = datetime.now(tz=ZoneInfo("America/New_York"))
+    return min(dt.day, constants.MAX_DAY)
```
Comment (proegssilb) [inline bot.py:48]: Is there a reason you didn't import `today()` from `lib.py`?

(Yes, I see the bug around `constants.MAX_DAY` vs. hard-coding 25, that should be fixed anyway)

File: constants.py [hunk 2]
```diff
@@ -1,17 +1,19 @@
...
+    description=f"""
 **help** - Send this message
 **info** - Some useful information about benchmarking
 **best _[day]_ _[part]_** - Best times so far for a day
 **submit _[day]_ _[part]_ <attachment>** - Benchmark attached code
 
 If [_day_] and/or [_part_] is ommited, they are assumed to be today and part 1
 
-Message <@{}> for any questions""",
+Message <@{config.settings.discord.owner_id}> for any questions""",
```
Comment (proegssilb) [inline constants.py:16]: A change like this, but much more thorough, is in https://github.com/proegssilb/ferris-elf/pull/53 . Can we undo the changes in this file to minimize merge conflicts?

File: bot.py [hunk 3]
```diff
@@ -1,107 +1,126 @@
...
+        await ctx.reply(embed=constants.HELP_REPLY)
+
+    @commands.hybrid_command()
+    async def info(self, ctx: commands.Context):
+        await ctx.reply(embed=constants.INFO_REPLY)
+
+
+async def prefix(dbot: commands.Bot, message: discord.Message) -> typing.List[str]:
+    # TODO: guild-specific prefixes
+    return ["aoc ", "aoc", f'<@!{dbot.user.id}> ', f'<@{dbot.user.id}> ', f'<@!{dbot.user.id}>', f'<@{dbot.user.id}>']
```
Comment (proegssilb) [inline bot.py:123]: Why are there duplicates in this list?

File: lib.py [hunk 4]
```diff
@@ -48,34 +49,40 @@ async def benchmark(submit_msg: SubmitMessage):
...
-        await msg.reply(f"Unhandled exception while benchmarking day {day}, part {part}.")
+    except Exception as e:
+        logger.exception(f"Unhandled exception while benchmarking day {day}, part {part}.")
+        with io.BytesIO() as buf:
+            buf.write(bytes(''.join(
+                traceback.format_exception(e)), encoding='utf8'))
+            buf.seek(0)
+            errtxt = (f"Unhandled exception while benchmarking day {day}, part {part}: `{get_full_class_name(e)}`: "
+                      f"`{e}`")[:2000]
+            await ctx.reply(errtxt, file=discord.File(buf, filename="traceback.txt"))
```
Comment (proegssilb) [inline lib.py:85]: There's an open question whether we should do this or not. People can use `eprintln!` and then panic to leak the input file, which could be good or bad.

**Review by proegssilb — COMMENTED** [review]
Couple of questions, but this looks like it's on a good track. 

I haven't looked at the error handler yet, that's going to need more brainpower than I can commit right now.

File: bot.py (hunk 1 above)
Comment (reticivis-net) [reply bot.py:48]: did not know that there was this function in lib.py, i was just copying the function that existed in messages.py. will fix

File: constants.py (hunk 2 above)
Comment (reticivis-net) [reply constants.py:16]: fair, can do

File: bot.py (hunk 3 above)
Comment (reticivis-net) [reply bot.py:123]: they're not duplicates, one has a space after and the other doesn't. a bit ugly to hardcode like this but id rather not generate it for every message sent

File: lib.py (hunk 4 above)
Comment (reticivis-net) [reply lib.py:85]: ah
the existing ferris elf bot does give some traceback so i thought we would do the same, im not sure how that one filters for eprintln panics

File: error_handler.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+class ErrorHandlerCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot: commands.Bot = bot
+        self.antispambucket = {}
+
+    @commands.Cog.listener()
+    async def on_command_error(self, ctx: commands.Context, commanderror: commands.CommandError):
+        async def dmauthor(*args, **kwargs):
+            try:
+                return await ctx.reply(*args, **kwargs)
```
Comment (proegssilb) [inline error_handler.py:34]: Does this actually use a direct message consistently? If not, does the code need to change, or does the method name need to change?

File: error_handler.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                    return await dmauthor(msg, file=file, embed=embed)
+
+        async def logandreply(message):
+            if ctx.guild:
+                ch = f"channel #{ctx.channel.name} ({ctx.channel.id}) in server {ctx.guild} ({ctx.guild.id})"
+            else:
+                ch = "DMs"
+            logger.info(f"Command '{ctx.message.content}' by "
+                        f"@{ctx.message.author.name}#{ctx.message.author.discriminator} ({ctx.message.author.id}) "
+                        f"in {ch} failed due to {message}.")
```
Comment (proegssilb) [inline error_handler.py:63]: Can we clean this up? I think it'd benefit from not using f-strings. Plus, discriminators don't exist anymore.

Comment (proegssilb) [reply error_handler.py:63]: ...looking at how this is used, should it be using `logger.exception()`?

File: bot.py [hunk 7]
```diff
@@ -1,107 +1,120 @@
...
+    @commands.hybrid_command()
+    async def help(self, ctx: commands.Context):
+        await ctx.reply(embed=constants.HELP_REPLY)
+
+    @commands.hybrid_command()
+    async def info(self, ctx: commands.Context):
+        await ctx.reply(embed=constants.INFO_REPLY)
+
+
+async def prefix(dbot: commands.Bot, message: discord.Message) -> typing.List[str]:
```
Comment (ultrabear) [inline bot.py:115]: we use python 3.11 in this house (currently)
so we can use `list[str]` here, no need to use an ugly import
also it is common practice to import things from typing directly, since they are pseudo builtin to typed python

File: bot.py [hunk 8]
```diff
@@ -1,107 +1,120 @@
...
-import constants
+from database import Database
+from error_handler import ErrorHandlerCog
 
 logger = logging.getLogger(__name__)
 
 
-class MyBot(discord.Client):
+class MyBot(commands.Bot):
     def __init__(self, *args, **kwargs):
```
Comment (ultrabear) [inline bot.py:21]: we should probably add `__slots__`, its a perf benefit but it also helps typecheckers assert what can be a member of the class
```suggestion
    __slots__ = ("queue", "db")
    
    def __init__(self, *args, **kwargs):
```

File: bot.py [hunk 9]
```diff
@@ -1,107 +1,120 @@
...
+                await lib.benchmark(*submit_msg)
                 self.queue.task_done()
             except Exception:
                 logger.exception("Error while processing submission.")
 
-    async def handle_best(self, msg):
+
+# if i don't use a cog, the functions would need to be in __name__ == __main__
+class Commands(commands.Cog):
+    def __init__(self, sbot: MyBot):
```
Comment (ultrabear) [inline bot.py:46]: Same `__slots__` comment here
```suggestion
    __slots__ = ("bot",)

    def __init__(self, sbot: MyBot):
```

File: bot.py [hunk 10]
```diff
@@ -1,107 +1,120 @@
...
+    @commands.command()
+    @commands.is_owner()
+    async def sync(self, ctx: commands.Context):
+        # sync slash commands
+        # this is a separate command because rate limits
+        await self.bot.tree.sync()
+        await ctx.reply("synced!")
+
+    # intentionally did not use typing.Optional because dpy treats it differently and i dont want that behavior
+    @commands.hybrid_command(aliases=["aoc", "lb"])
```
Comment (ultrabear) [inline bot.py:58]: personally i think this should just be `lb`, aoc is like a prefix, best is vague too

File: bot.py [hunk 11]
```diff
@@ -1,107 +1,120 @@
...
+        )
+        # using a tuple is probably the most readable but shut
+        self.bot.queue.put_nowait((ctx, day, part, await code.read()))
+        await ctx.reply(
+            f"Your submission for day {day} part {part} has been queued. "
+            f"There are {self.bot.queue.qsize()} submissions in the queue."
+        )
+
+    @commands.hybrid_command()
+    async def help(self, ctx: commands.Context):
```
Comment (ultrabear) [inline bot.py:107]: ik this PR is scope limited, but just leaving a note to remember migrate to the builtin help that dpy Bot provides

File: bot.py [hunk 12]
```diff
@@ -122,5 +135,13 @@ async def on_message(self, msg: discord.Message):
...
         sys.exit(1)
-    bot = MyBot(intents=intents)
+    bot = MyBot(
+        intents=intents,
+        command_prefix=prefix,
+        case_insensitive=True,
+        # dpy comes with a help command, we need to remove it
+        help_command=None,
+        # disallows the bot from mentioning things it shouldn't, just in case
+        allowed_mentions=discord.AllowedMentions(everyone=False, users=True, roles=False, replied_user=True)
```
Comment (ultrabear) [inline bot.py:145]: For anyone new to dpy, these are not ulimit style, i/e, any time we want to override this for whatever reason we can, but having it off by default is safe

File: error_handler.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+import traceback
+import urllib
+
+import discord
+from discord.ext import commands
+
+logger = logging.getLogger(__name__)
+
+
+def get_full_class_name(obj):
```
Comment (ultrabear) [inline error_handler.py:13]: This function is very confusing to look at, i think it needs a doc comment for clarity

File: error_handler.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    module = obj.__class__.__module__
+    if module is None or module == str.__class__.__module__:
+        return obj.__class__.__name__
+    return module + '.' + obj.__class__.__name__
+
+class NonBugError(Exception):
```
Comment (ultrabear) [inline error_handler.py:19]: tbh I dont like this name, but we can put off bikeshedding it for later

File: error_handler.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+            if ctx.interaction:
+                if ctx.interaction.response.is_done():
+                    return await ctx.interaction.edit_original_response(content=msg,
+                                                                        attachments=[file] if file else None,
+                                                                        embed=embed)
+                else:
+                    return await ctx.reply(msg, file=file, embed=embed)
+            else:
+                try:
+                    if ctx.guild and not ctx.channel.permissions_for(ctx.me).send_messages:
```
Comment (ultrabear) [inline error_handler.py:48]: this is a TOCTOU bug, since permissions can change between us checking and sending the reply, we should attempt to send the message always and catch errors it gives

File: error_handler.py (hunk 6 above)
Comment (ultrabear) [reply error_handler.py:63]: > discriminators don't exist anymore.

this is (unfortunately?) false, bots still have them, and so do old unmigrated accounts, code still needs to account for it

File: error_handler.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+            logger.info(f"Command '{ctx.message.content}' by "
+                        f"@{ctx.message.author.name}#{ctx.message.author.discriminator} ({ctx.message.author.id}) "
+                        f"in {ch} failed due to {message}.")
+            await reply(message)
+
+        errorstring = str(commanderror)
+        if isinstance(commanderror, discord.Forbidden):
+            await dmauthor(f"I don't have permissions to send messages in that channel.")
+            logger.info(commanderror)
+        if isinstance(commanderror, discord.ext.commands.errors.CommandNotFound):
```
Comment (ultrabear) [inline error_handler.py:70]: can we replace this with match? the isinstance spam is hard to look at
and also `import discord.ext.commands.errors as boterror` would be nice

**Review by ultrabear — CHANGES_REQUESTED** [review]
Mostly good, a few things should probably be fixed now though

File: error_handler.py (hunk 16 above)
Comment (proegssilb) [reply error_handler.py:70]: I thought about this. I can't find clear info on whether or not Python match statements return values or not (I'm guessing they don't), but it'd be nice if there were a way to extract out the repeated `await logandreply(err)` calls as well. I got distracted by not being quite able to figure out how to make it all work.

File: error_handler.py (hunk 6 above)
Comment (ultrabear) [reply error_handler.py:63]: however in this case the default user repr accounts for it so we can just pass ctx.message.author to fmt

File: error_handler.py (hunk 16 above)
Comment (reticivis-net) [reply error_handler.py:70]: does python match work with isinstance? I thought it didn't

File: error_handler.py (hunk 6 above)
Comment (reticivis-net) [reply error_handler.py:63]: > Can we clean this up? I think it'd benefit from not using f-strings. Plus, discriminators don't exist anymore.

for me, f-strings are way easier to read, especially when there's many parameters. This code was directly lifted from another bot of mine that I haven't updated in a while. 
Also I believe i used logger.info because a full logger.exception traceback is not necessary for non-errors like invalid commands

File: error_handler.py (hunk 5 above)
Comment (reticivis-net) [reply error_handler.py:34]: code needs to change, not entirely sure what's going on there

File: bot.py (hunk 10 above)
Comment (reticivis-net) [reply bot.py:62]: I dont think having extra aliases hurts anything

File: bot.py (hunk 11 above)
Comment (reticivis-net) [reply bot.py:111]: builtin dpy help is limited and I generally prefer to use my own rather than hacking dpy's. i can adapt a much more complex help command based on docstrings i used in another bot, but i think that's overkill for a bot with 4 commands

File: error_handler.py (hunk 16 above)
Comment (proegssilb) [reply error_handler.py:72]: Seemingly, but I'm not 100% clear how exactly that all works. https://peps.python.org/pep-0636/#adding-a-ui-matching-objects

Comment (ultrabear) [reply error_handler.py]: > does python match work with isinstance? I thought it didn't

matches primary use is replacing isinstance with just raw type matching, ~~copied off~~ similar to rust

Comment (proegssilb) [reply error_handler.py]: https://www.geeksforgeeks.org/python-match-case-statement/ clarifies a bit:

> We can also use [Python classes](https://www.geeksforgeeks.org/python-classes-and-objects/) to match using the Python match statements. This will make the code very much clean,  neat, and more importantly easily readable.
>
> The first thought after seeing the match statement would be that the Programmer and Person inside the match statement would create an instance of the respective classes, but it is not the case. It will check for the instance with the given values. So the case Programmer(“Om”, language = “Python”, framework = “Django”)  will not create an instance but will check that the object passed whose name is instance is a  real instance of the Programmer class, then it will check for the name that is  “Om”, language that is “Python” and then framework that is “Django”. This way we can check for the values in the instance of any class.

Check the section "Use of Python Class in match statement" for more, but the fact that it's using construction syntax for an attribute list is at least a way of explaining what's going on, even if it's a bit surprising.

The full rules, given at https://docs.python.org/3/reference/compound_stmts.html#class-patterns , are rather involved.

Comment (reticivis-net) [reply error_handler.py]: Already updated the code to use a match statement

File: bot.py [hunk 17]
```diff
@@ -115,6 +115,10 @@ async def help(self, ctx: commands.Context):
     async def info(self, ctx: commands.Context):
         await ctx.reply(embed=constants.INFO_REPLY)
 
+    @commands.hybrid_command()
+    async def error(self, ctx:commands.Context):
+        raise Exception("peepee")
+
```
Comment (proegssilb) [inline bot.py:121]: I'm sorry, I'm going to need an explanation on this one.

File: error_handler.py [hunk 18]
```diff
@@ -40,21 +41,20 @@ async def dmauthor(*args, **kwargs):
...
-                                                                        embed=embed)
+            try:
+                if ctx.interaction and ctx.interaction.response.is_done():
+                    return await ctx.interaction.edit_original_response(
+                        content=msg,
+                        attachments=[file] if file else None,
+                        embed=embed
+                    )
                 else:
                     return await ctx.reply(msg, file=file, embed=embed)
```
Comment (proegssilb) [inline error_handler.py:52]: Is avoiding the race condition better, or is the cost from the extra API call too much for the race condition?

(Please check the formatting before replying, I'm probably off by a space somewhere)

```suggestion
            try:
                return await ctx.interaction.edit_original_response(
                        content=msg,
                        attachments=[file] if file else None,
                        embed=embed
                    )
            except discord.NotFound:
                return await ctx.reply(msg, file=file, embed=embed)
```

**Review by proegssilb — COMMENTED** [review]
Two points of clarification from me. Sorry to drag this out!

File: bot.py (hunk 17 above)
Comment (reticivis-net) [reply bot.py:121]: THAT WAS TESTING CODE I DID **NOT** MEAN TO COMMIT THAT OH MY GOD

File: error_handler.py (hunk 18 above)
Comment (reticivis-net) [reply error_handler.py]: I don't think there is a race condition. Once an interaction is done, it can't be "undone", it's a permanent state.

Comment (proegssilb) [reply error_handler.py]: I was thinking "response was in-progress at the time of check, but then finished while making the call to `ctx.reply()`, leading to a double-reply". You'd know better than me, though. If that's not a risk, then I'm happy.

**Review by ultrabear — APPROVED** [review]
Changes LGTM
